RenderSystems/GLSupport/src/EGL/Wayland/OgreWaylandEGLWindow.cpp

```diff
@@ -0,0 +1,333 @@
+// This file is part of the OGRE project.
+// It is subject to the license terms in the LICENSE file found in the top-level directory
+// of this distribution and at https://www.ogre3d.org/licensing.
+// SPDX-License-Identifier: MIT
+
+#include "OgreException.h"
+#include "OgreLogManager.h"
+#include "OgreRoot.h"
+#include "OgreStringConverter.h"
+#include "OgreViewport.h"
+
+#include "OgreWaylandEGLSupport.h"
+#include "OgreWaylandEGLWindow.h"
+
+#include <thread>
+#include <chrono>
+
+namespace Ogre
+{
+
+WaylandEGLWindow::WaylandEGLWindow(WaylandEGLSupport* glsupport) : EGLWindow(glsupport)
+{
+    mGLSupport = glsupport;
+    mNativeDisplay = nullptr; //glsupport->getNativeDisplay();
+}
+
+WaylandEGLWindow::~WaylandEGLWindow()
+{
+    //mNativeDisplay = mGLSupport->getNativeDisplay();
+
+    if (mWindow && mIsTopLevel)
+    {
+        if (!mIsExternal)
+            wl_egl_window_destroy(mWindow);
+    }
+    mWindow = nullptr;
+}
+
+void WaylandEGLWindow::getCustomAttribute(const String& name, void* pData)
+{
+    EGLWindow::getCustomAttribute(name, pData);
+    // #TODO add support for wayland-related variables;
+    if (name == "WL_DISPLAY")
+    {
+        *static_cast<NativeDisplayType*>(pData) = mGLSupport->getNativeDisplay();
+        return;
+    }
+    else if (name == "WL_WINDOW")
+    {
+        *static_cast<NativeWindowType*>(pData) = mWindow;
+        return;
+    }
+}
+
+void WaylandEGLWindow::initNativeCreatedWindow(const NameValuePairList* miscParams)
+{
+    //mNativeDisplay = mGLSupport->getNativeDisplay();
+
+    if (miscParams)
+    {
+        NameValuePairList::const_iterator opt;
+        NameValuePairList::const_iterator end = miscParams->end();
+
+        if ((opt = miscParams->find("externalDisplay")) != end)
+        {
+          StringVector tokens = StringUtil::split(opt->second, " :");
+
+          auto tmp = (wl_display*)StringConverter::parseSizeT(tokens[0]);
+          mNativeDisplay = tmp;
+          mGLSupport->setNativeDisplay(mNativeDisplay);
+
+        }
+        if ((opt = miscParams->find("parentWindowHandle")) != end ||
+            (opt = miscParams->find("externalWindowHandle")) != end)
+        {
+            StringVector tokens = StringUtil::split(opt->second, " :");
+
+            // Qt: This casts QWindow to wl_egl_window, which may not work..
+            mWindow = (wl_egl_window*)StringConverter::parseSizeT(tokens[0]);
+            mIsExternal = true;
+            mIsTopLevel = true;
+        }
+        if ((opt = miscParams->find("externalSurface")) != end)
+        {
+            StringVector tokens = StringUtil::split(opt->second, " :");
+
+            mGLSupport->mWlSurface = (wl_surface*)StringConverter::parseSizeT(tokens[0]);
+        }
+    }
+    mNativeDisplay = mGLSupport->getNativeDisplay();
+}
+
+void WaylandEGLWindow::createNativeWindow(uint& width, uint& height, String& title)
```

Comment:
I would just assume that the window is provided by "parentWindowHandle" above. This way we dont have to deal with the compositor at all here.

Reply:
Agree.